test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java

```diff
@@ -0,0 +1,139 @@
+
+    private static final AtomicBoolean doInBackgroundStarted = new AtomicBoolean(false);
+    private static final AtomicBoolean doInBackgroundFinished = new AtomicBoolean(false);
+    private static final CountDownLatch doneLatch = new CountDownLatch(1);
+
+    public static void main(String[] args) throws InterruptedException {
+        SwingWorker<String, String> worker = new SwingWorker<>() {
+            @Override
+            protected String doInBackground() throws Exception {
+                try {
+                    while (!Thread.currentThread().isInterrupted()) {
+                        System.out.println("Working...");
+                        Thread.sleep(WAIT_TIME);
+                    }
+                } catch (InterruptedException ex) {
+                    System.out.println("Got interrupted!");
+                }
+
+                try {
+                    doInBackgroundStarted.set(true);
+                    System.out.println("Cleaning up");
+                    Thread.sleep(CLEANUP_TIME);
+                    System.out.println("Done cleaning");
+                    doInBackgroundFinished.set(true);
+                } catch (InterruptedException ex) {
+                    System.out.println("Got interrupted second time!");
+                }
+
+                return null;
+            }
+
+            @Override
+            protected void done() {
+                if (!doInBackgroundFinished.get()) {
+                    throw new RuntimeException("done called before " +
+                                               "doInBackground is finished");
+                }
+                System.out.println("Done");
+                doneLatch.countDown();
+            }
+        };
+
+        worker.addPropertyChangeListener(
+            new PropertyChangeListener() {
+                public void propertyChange(PropertyChangeEvent evt) {
+                    // Before doInBackground method is invoked,
+                    // SwingWorker notifies PropertyChangeListeners about the
+                    // property change to StateValue.STARTED
+                    if (doInBackgroundFinished.get()
+                        && worker.getState() == SwingWorker.StateValue.STARTED) {
+                        throw new RuntimeException(
+                               "PropertyChangeListeners called with " +
+                               "state STARTED after doInBackground is invoked");
+                    }
+
+                    // Ensure the STARTED state is notified
+                    // before doInBackground started running
+                    if (doInBackgroundStarted.get()
+                        && worker.getState() == SwingWorker.StateValue.STARTED) {
+                        throw new RuntimeException(
+                              "PropertyChangeListeners called with " +
+                               "state STARTED before doInBackground is finished");
+                    }
```

Comment:
The comments, conditions and error messages do not match.

In the first case, it should be `doInBackgroundStarted` instead of `doInBackgroundFinished`; in the second — `doInBackgroundFinished` correspondingly.

Reply:
Test is updated to test the order